packages/react/toggle-button/src/ToggleButton.tsx

```diff
@@ -57,12 +247,138 @@ const ToggleButton = forwardRefWithAs<typeof DEFAULT_TAG, ToggleButtonOwnProps>(
   }
 );
 
-ToggleButton.displayName = NAME;
+ToggleButton.displayName = BUTTON_NAME;
 
 const Root = ToggleButton;
+const Group = ToggleButtonGroup;
+const GroupExclusive = ToggleButtonGroupExclusive;
 
-export {
-  ToggleButton,
-  //
-  Root,
-};
+export { ToggleButton, ToggleButtonGroup, ToggleButtonGroupExclusive, Root, Group, GroupExclusive };
+
+function useConflictingContextError(
+  standardGroupContext: ToggleButtonGroupContextValue | null,
+  exclusiveGroupContext: ToggleButtonGroupExclusiveContextValue | null
+) {
+  if (exclusiveGroupContext && standardGroupContext) {
+    throw Error(
+      `A ${BUTTON_NAME} was used in both ${GROUP_EXC_NAME} and ${GROUP_NAME} components. ${BUTTON_NAME} can be used in either ${GROUP_EXC_NAME} or ${GROUP_NAME}, but not both.`
+    );
+  }
+}
+
+function useButtonStateInGroupWarning({
+  toggledProp,
+  defaultToggled,
+  groupContext,
+  groupName,
+}: {
+  toggledProp: boolean | undefined;
+  defaultToggled: boolean | undefined;
+  groupContext: ToggleButtonGroupContextValue | ToggleButtonGroupExclusiveContextValue | null;
+  groupName: string | null;
+}) {
+  const isControlled = useIsControlled(toggledProp);
+  if (__DEV__) {
+    // eslint-disable-next-line react-hooks/rules-of-hooks
+    React.useEffect(() => {
+      const propName = isControlled ? '`toggled`' : '`defaultToggled`';
+      const toggled = isControlled ? toggledProp : defaultToggled;
+      warning(
+        !(groupContext && toggled !== undefined),
+        `A ${BUTTON_NAME} with an explicit ${propName} prop was used inside of a ${groupName}. When ${BUTTON_NAME} is used inside of a ${groupName}, the ${groupName} component is responsible for managing the toggled state of its nested ${BUTTON_NAME} components. Either remove the ${propName} from ${BUTTON_NAME} or remove the ${groupName} component if the ${BUTTON_NAME} is unrelated to other surrounding buttons.`
+      );
+    }, [groupContext, toggledProp, defaultToggled, groupName, isControlled]);
+  }
+}
+
+function stringArraysAreEqual(arr1: string[] | undefined, arr2: string[] | undefined) {
+  if (arr1 === undefined && arr2 === undefined) return true;
+  if (arr1 === undefined) return false;
+  if (arr2 === undefined) return false;
+  return JSON.stringify(arr1.sort()) === JSON.stringify(arr2.sort());
+}
+
+// NOTE: I changed this implementation a tiny bit to improve typing somewhat. I think it's a little
+// nicer not to include `undefined` in the typing here and always provide a defaultProp value
+// whether or not its passed by the consumer. If undefined is an acceptable type it should be added
```

Comment:
This has nothing to do with the consumer passing `defaultValue`, we use it internally and always pass a declared variable here AFAICT. If the consumer doesn't pass it, it's just `undefined`.

Reply:
When I say "consumer" here, I mean us. Whoever consumes this hook (us) has to pass a `defaultValue` to solve a problem with the hook's types. You're moving the problem onto the consumption end to solve. It would be better to solve the problem in the hook because it is the hook's problem and not the code consuming it.

Really bad at explaining things as always but hopefully that makes more sense 😅

Reply:
I'm gonna try help take a look at this one. I'm guessing you did it so that `const [value, setValue]` gives you a `[string, ...]` type instead of `[string | undefined, ...]` ?

Reply:
Yes, that's correct. I think it's a little nicer if we do either provide a `defaultValue` in every case unless we want to explicitly allow `undefined`. It's a little more to do when using the function but more explicit is nice IMO, and easier to work around if we don't want or need `undefined` to be a valid type.

Reply:
Ahhhh I've had a closer look at this and I think I see what's going on now! 🙈 

My misunderstanding was that I thought you wanted the hook to always give back `[string, ...]` but no, we don't always want that because it is possible for `prop` **and** `defaultProp` to be `undefined` e.g. `<ToggleButtonGroup />. 

In that case, it should give you back `[string | undefined, ...]` so the **hook** is reporting correctly but in **your particular case** you want to say "always fallback to a value if both are undefined" so it would give you `[string, ...]` 🎉  I'm with you now!! Haha sorry 😅 

For that case, I have just done the following in previous components:

```jsx
const [checked = false, setChecked] = useControlledState({
  prop: checkedProp,
  defaultProp: defaultChecked,
});
```

That should work with our existing hook without any type changes I believe and tells me that both of those could be undefined so you want to fallback to `false` in that case.

Reply:
@jjenzz In that case `setChecked` would still be typed as `setChecked(val: boolean | undefined): void`. This workaround still presents annoying issues like this, since the value in our setter should never be undefined.

![image](https://user-images.githubusercontent.com/3082153/104644651-e1340a80-5662-11eb-8b37-5b2544cc5bc5.png)

Reply:
I've actually noticed this is an issue when something like `setState(prevState => [...prevState, newValue])` is called, we always have to do `setState((prevState = []) => ....` which is annoying.

I'll take a look at solving all this stuff separately across our codebase at some point. Thanks for being patient with me on this one 🙈